File: Leetcode/Dynamic Programming/312. Burst Balloons.cpp
class Solution {
private:
    int N;
    vector <vector <int>> DP;
    int helper(int L, int R, vector <int> &nums) {
        if (L > R) return 0;
        if (DP[L][R] != -1) return DP[L][R];
        for (int i = L; i <= R; i++) {
            DP[L][R] = max(DP[L][R], helper(L, i - 1, nums) + helper(i + 1, R, nums) + nums[i] * ((L > 0) ? nums[L - 1] : 1) * ((R + 1 < N) ? nums[R + 1] : 1));
        }
        return DP[L][R];
    }
public:
    int maxCoins(vector<int>& nums) {
        N = nums.size();
        DP = vector <vector <int>> (N, vector <int> (N, -1));
        return helper(0, nums.size() - 1, nums);
    }
};

File: Leetcode/Dynamic Programming/312. Burst Balloons.java
class Solution {
    private int[][] DP;
    private int N;
    private int helper(int L, int R, int[] nums) {
        if (L > R) return 0;
        if (DP[L][R] != -1) return DP[L][R];
        for (int i = L; i <= R; i++) {
            DP[L][R] = Math.max(DP[L][R], helper(L, i - 1, nums) + helper(i + 1, R, nums) + nums[i] * ((L > 0) ? nums[L - 1] : 1) * ((R + 1 < N) ? nums[R + 1] : 1));
        }
        return DP[L][R];
    }
    public int maxCoins(int[] nums) {
        N = nums.length;
        DP = new int[N][N];
        for (int[] row: DP) {
            Arrays.fill(row, -1);
        }
        return helper(0, N - 1, nums);
    }
}